File: section2/115.不同的子序列.py
#
# @lc app=leetcode.cn id=115 lang=python3
#
# [115] 不同的子序列
#

# @lc code=start
class Solution:
    def num_distinct_recursive(self, s, left, t, right, mems):
        key = f'{left}-{right}'
        if key not in mems:
            if right == len(t):
                num = 1
            elif left >= len(s) or right >= len(t):
                num = 0
            else:
                num = self.num_distinct_recursive(s, left + 1, t, right, mems)
                if s[left] == t[right]:
                    num += self.num_distinct_recursive(s, left + 1, t, right + 1, mems)

            mems[key] = num

        return mems[key]

    def numDistinct(self, s: str, t: str) -> int:
        return self.num_distinct_recursive(s, 0, t, 0, {})

    def test(self):
        print(self.numDistinct('rabbbit', 'rabbit') == 3)
        print(self.numDistinct('babgbag', 'bag') == 5)


# Solution().test()

# @lc code=end

File: section2/section/115.不同的子序列.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=115 lang=golang
 *
 * [115] 不同的子序列
 */

// @lc code=start
func numDistinct(s string, t string) int {
	nums := make([][]int, len(s)+1)
	for index := range nums {
		nums[index] = make([]int, len(t)+1)
		nums[index][0] = 1
	}

	for i := 1; i <= len(s); i++ {
		for j := 1; j <= i && j <= len(t); j++ {
			nums[i][j] = nums[i-1][j]
			if s[i-1] == t[j-1] {
				nums[i][j] += nums[i-1][j-1]
			}
			fmt.Printf("[%d, %d]=[%s, %s]=%d\n", i, j, s[0:i], t[0:j], nums[i][j])
		}
	}

	return nums[len(s)][len(t)]
}

func Test115() {
	fmt.Println(numDistinct("rabbbit", "rabbit"))
	fmt.Println(numDistinct("bab", "b"))
	fmt.Println(numDistinct("babgbag", "bag"))
}

// @lc code=end
